### PR TITLE
CA-226000: Old patches are not removed on Host Upgrade

### PR DESCRIPTION
**Pull request by fillzero**
Remove any pool_patch objects that don't have a corresponding
pool_update object

Signed-off-by: Liang Dai <liang.dai1@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_pool_update.ml [hunk 1]
```diff
@@ -445,7 +445,16 @@ let resync_host ~__context ~host =
       ) update_refs;
     Create_misc.create_updates_requiring_reboot_info ~__context ~host
   end
-  else Db.Host.set_updates ~__context ~self:host ~value:[]
+  else Db.Host.set_updates ~__context ~self:host ~value:[];
+
+  (* Remove any pool_patch objects that don't have a corresponding pool_update object *)
+  let pool_patchs = Db.Pool_patch.get_all ~__context in
```
Comment (lindig) [inline ocaml/xapi/xapi_pool_update.ml]: This should be `pool_patches` and likewise below. Removing unused patches can be expressed more concisely with `List.filter`:


    Db.Pool_patch.get_all ~__context
    |> List.map (fun self -> Db.Pool_patch.get_pool_update ~__context ~self)
    |> List.filter Ref.null
    |> List.iter (fun self -> Db.Pool_patch.destroy ~__context ~self)

Comment (fillzero) [inline ocaml/xapi/xapi_pool_update.ml]: Hi @lindig ,
Thanks for your comments.
Updated the commit.

**Comment by lindig** [comment]
This looks good.

**Comment by lindig** [comment]
This has the `mustfix` label. Should this be merged for the Ely release?

**Comment by lindig** [comment]
The code change removes the objects from the database. What about files in the file system belonging to the patches? Or is this not a concern?

**Comment by jonludlam** [comment]
The files won't be an issue since this is mainly addressing stale state left behind after an upgrade.

However, I do think we ought to do the symmetric check too, of cleaning updates that don't have a corresponding patch record. That way we can't be left in an inconsistent state.

**Comment by fillzero** [comment]
@lindig  @jonludlam  Thanks a lot.
Updated the commit.

**Comment by lindig** [comment]
This looks good.

**Pull request merged**